### PR TITLE
[SDP-1317] Add MessageDispatcher to SDP 

### PR DESCRIPTION
**Pull request by marwen-abid**
### What
* Add `MessageDispatcher` that dispatches messages to a specific `MessageChannel`
* Currently supported `MessageChannels` are `SMS` and `EMAIL`. In the future we can add other channels like `WHATSAPP` 

### Why
Prepare for dynamically dispatching messages to a channel depending on the organization configuration. 

### Known limitations

[TODO or N/A]

#### PR Structure

* [ ] This PR has a reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR title and description are clear enough for anyone to review it.
* [ ] This PR does not mix refactoring changes with feature changes (split into two PRs otherwise).

#### Thoroughness

* [ ] This PR adds tests for the new functionality or fixes.
* [ ] This PR contains the link to the Jira ticket it addresses.

#### Configs and Secrets

* [ ] No new CONFIG variables are required -OR- the new required ones were added to the helmchart's [`values.yaml`] file.
* [ ] No new CONFIG variables are required -OR- the new required ones were added to the deployments ([`pr-preview`], [`dev`], [`demo`], `prd`).
* [ ] No new SECRETS variables are required -OR- the new required ones were mentioned in the helmchart's [`values.yaml`] file.
* [ ] No new SECRETS variables are required -OR- the new required ones were added to the deployments ([`pr-preview secrets`], [`dev secrets`], [`demo secrets`], `prd secrets`).

#### Release

* [ ] This is not a breaking change.
* [ ] **This is ready for production.**. If your PR is not ready for production, please consider opening additional complementary PRs using this one as the base. Only merge this into `develop` or `main` after it's ready for production!

#### Deployment

* [ ] Does the deployment work after merging?

[`values.yaml`]: ../helmchart/sdp/values.yaml
[`pr-preview`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/stellar-disbursement-platform/backend-helm-values
[`dev`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/backend-helm-values
[`demo`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-backend-helm-values
[`pr-preview secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/externalsecrets-common-previews.yaml#L241-L346
[`dev secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/stellar-disbursement-platform-externalsecrets.yaml
[`demo secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-sdp-externalsecrets.yaml

### CODE REVIEW

**Review by marcelosalloum — APPROVED** [review]
This is looking great!

Plz feel free to merge it after addressing the comments below:

File: internal/message/messenger_client.go [hunk 1]
```diff
@@ -1,5 +1,6 @@
 package message
 
+//go:generate mockery --name=MessengerClient  --case=underscore --structname=MessengerClientMock --inpackage --filename=mocks.go
```
[the remark this answers is not in the document]
Comment (marwen-abid) [reply internal/message/messenger_client.go:3]: I did have a circular dependency here which is why I switched it!

File: internal/message/message_dispatcher.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		return fmt.Errorf("getting client for channel: %w", err)
+	}
+
+	return client.SendMessage(message)
+}
+
+func (d *MessageDispatcher) GetClient(channel MessageChannel) (MessengerClient, error) {
+	client, ok := d.clients[channel]
+	if !ok {
+		return nil, fmt.Errorf("no client registered for channel: %s", channel)
```
[the remark this answers is not in the document]
Comment (marwen-abid) [reply internal/message/message_dispatcher.go]: Addressed in f7fb37107dc583a61705a2da22b4ecb5556d1b93

File: internal/message/mocks/message_dispatcher_interface.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
+// Code generated by mockery v2.40.1. DO NOT EDIT.
+
+package mocks
+
+import (
+	context "context"
+
+	message "github.com/stellar/stellar-disbursement-platform-backend/internal/message"
```
[the remark this answers is not in the document]
Comment (marwen-abid) [reply internal/message/mocks/message_dispatcher_interface.go]: Addressed in f7fb37107dc583a61705a2da22b4ecb5556d1b93

File: cmd/serve_test.go [hunk 4]
```diff
@@ -190,11 +190,18 @@ func Test_serve(t *testing.T) {
 	require.NoError(t, err)
 	serveOpts.CrashTrackerClient = crashTrackerClient
 
-	messengerClient, err := di.NewEmailClient(di.EmailClientOptions{EmailType: message.MessengerTypeDryRun})
+	emailOpts := di.EmailClientOptions{EmailType: message.MessengerTypeDryRun}
+	messengerClient, err := di.NewEmailClient(emailOpts)
```
[the remark this answers is not in the document]
Comment (marwen-abid) [reply cmd/serve_test.go]: Fixed in 53b4a22a2b748ebcb97f3b2da2799046e509d3d7

**Pull request merged**